articles/tcp-node/index.md

```diff
@@ -0,0 +1,237 @@
+const net = require("net"); // import net
+
+// create the server
+let server = net.createServer(connection => {
+    // run all of this when a client connects
+    console.log("new connection");
+
+    connection.on("data", data => {
+        // run this when data is received
+        console.log(data.toString()); // prints the data
+    });
+});
+
+// ...
+```
+
+We want to do a little more than print out the data though. We want to an actual calculation. Now, we should implement the protocol in our server.
+
+```JavaScript
+//...
+    connection.on("data", data => {
+        // run this when data is received
+        if (data == undefined || data == null) {
+			      return;
+		    }
+
+		    const dataArgs = data.toString().split(" "); // splits the data into spaces
+
+	      if (dataArgs.length === 0) { // in case there is no command
+            connection.write("ERROR no data");
+            return; // prevents other code from running
+        }
+		    const command = dataArgs[0]; // gets the command
+        if (command === "ADD") { // add command
+            if (dataArgs.length !== 3) { // in case there aren't enough arguments
+                connection.write("ERROR incorrect number of arguments");
+                return;
+			      }
+
+            const op1 = parseInt(dataArgs[1]); // first number
+            const op2 = parseInt(dataArgs[2]); // second number
+			      const result = (op1 + op2).toString(); // result as a string
+
+			      if (result === "NaN") { // in case the inputs aren't numbers
+				        connection.write("ERROR invalid number");
+				        return;
+			      }
+
+			      connection.write("RESULT " + result);
+			      return; // end
+		    } else { // invalid command
+            connection.write("ERROR invalid command");
+            return;
+        }
+    });
+
+// ...
+```
+
+You can run this by typing `node server.js` in a command prompt or a terminal.
+
+### Client
+Now that the server is done, we need a program to send information and it needs to connect to the server.
+
+```JavaScript
+// client.js
+const net = require("net"); // import net
+const readline = require("readline").createInterface({
+        input: process.stdin,
+        output: process.stdout
+}); // this will be important later
+
+const options = {
+    port: 50000,
+    // if you want to connect to a different computer you can use host: "HOST"
+    // replace HOST with the IP address to connect to
+    // otherwise, run the server on the same computer as the client
+};
+
+let client = net.connect(options, () => {
+    console.log("connected!");
+});
+```
+
+The rest should be pretty easy. We'll ask for two numbers to be added together, and we'll send it to the server. Then, whenever we receive data back, we'll print the result and repeat the process.
+
+To [read data in Node.js](https://nodejs.dev/learn/accept-input-from-the-command-line-in-nodejs), we'll need the `readline` module.
+
+```JavaScript
+// client.js
+
+// ...
```

Comment:
@botahamec Do you need to import the readline module?

Reply:
@ninjaginja It was imported in the earlier code snippet